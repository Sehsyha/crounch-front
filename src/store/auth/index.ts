import { Module } from 'vuex';

import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { getters } from './getters';
import { actions } from './actions';
import { mutations } from './mutations';
import { RootState } from '..';

export interface AuthState {
  token: string;
  status: string;
}

export const state: AuthState = {
  token: localStorage.getItem(TOKEN_STORAGE_KEY) || '',
  status: '',
};

const namespaced: boolean = true;
export const authNamespace = { namespace: 'auth' };

export const auth: Module<AuthState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations,
};
