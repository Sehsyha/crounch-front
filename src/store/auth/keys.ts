export enum AuthActions {
  LOGIN = 'LOGIN',
  SIGNUP = 'SIGNUP',
  LOGOUT = 'LOGOUT',
}

export enum AuthGetters {
  IS_AUTHENTICATED = 'IS_AUTHENTICATED',

}

export enum AuthMutations {
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
}
