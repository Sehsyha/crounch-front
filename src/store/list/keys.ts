export enum ListKeys {
  CREATE = 'CREATE',
  CREATED = 'CREATED',
  GETOWNERS = 'GETOWNERS',

  ADD = 'ADD',
  SET = 'SET',
  RESET = 'RESET',
};
