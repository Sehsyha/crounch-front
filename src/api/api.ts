import { getAPIURL } from '@/utils/environment';
import { TOKEN_STORAGE_KEY } from '@/utils/constants';
import { FetchError } from '@/utils/error';
import { AuthModule } from '@/store/AuthModule';
import { getModule } from 'vuex-module-decorators';
import router from '@/router/router';

export interface FetchOptions {
  url: string;
  data?: object;
  method: string;
}

export function doFetch<T>(options: FetchOptions): Promise<T> {
  return fetch(getUrl(options.url), getFetchOptions(options))
    .then(res => {
      if (res.status >= 200 && res.status <= 299 && res.status !== 204) {
        return res.json()
      }
      if (res.status === 204) {
        return
      }
      if (res.status === 401) {
        const authModule = getModule(AuthModule)
        authModule.logoutAction().then(() => {
          router.push({ name: 'home' })
        })
      }

      return res.json()
        .then(body => { throw new FetchError('Erreur lors de la requÃªte', res.status, body) })
        .catch()
    })
}

function getFetchOptions({ data, method }: FetchOptions): object {
  let options: any = {
    method,
    headers: {
      Authorization: localStorage.getItem(TOKEN_STORAGE_KEY),
    },
  };

  if (method !== 'GET') {
    options.body = JSON.stringify(data);
  }

  return options;
}

function getUrl(url: string): string {
  return `${getAPIURL()}/${url}`;
}
