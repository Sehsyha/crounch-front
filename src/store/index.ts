import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { auth } from './auth';
import { ListModule } from './ListModule';
import { product } from './product'

Vue.use(Vuex);

export interface RootState { }

const store: StoreOptions<RootState> = {
  state: {},
  modules: {
    auth,
    ListModule,
    product,
  },
  strict: true,
};

export default new Vuex.Store<RootState>(store);
