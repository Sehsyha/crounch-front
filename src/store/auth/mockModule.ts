import { AuthKeys } from './keys';
import { Module } from 'vuex';
import { AuthState } from '.';
import { RootState } from '..';

const actions = {
  [AuthKeys.SIGNUP]: jest.fn(),
  [AuthKeys.LOGIN]: jest.fn(),
  [AuthKeys.LOGOUT]: jest.fn(),
};

const getters = {
  isAuthenticated: jest.fn(),
};

export function createAuthModuleMock(): Module<AuthState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    state: {
      token: '',
      status: '',
    },
  };
}
