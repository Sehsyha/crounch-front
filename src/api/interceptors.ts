import { api } from './api';

export const unauthorizedInterceptor = (logout: any) => (err: any): Promise<void> => {
  if (err.status === 401 && err.config && !err.config.__isRetryRequest) {
    logout();
  }
  return Promise.reject(err);
};

export function addUnauthorizedInterceptor(logout: any) {
  api.interceptors.response.use(undefined, unauthorizedInterceptor(logout));
}
