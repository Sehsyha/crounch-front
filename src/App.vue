<template>
  <div id="app">
    <div id="nav">
      <router-link v-if="!isAuthenticated" to="/">Home|</router-link>
      <router-link v-if="isAuthenticated" to="lists">Listes |</router-link>
      <router-link to="/about">About|</router-link>
      <router-link to="/version">Version</router-link>
    </div>
    <router-view />
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import 'vue-router';

import Component from 'vue-class-component';
import { Action, Getter } from 'vuex-class';
import { addUnauthorizedInterceptor } from './api/interceptors';
import { AuthKeys } from './store/auth/keys';
import { authNamespace } from './store/auth';

@Component
export default class App extends Vue {
  @Action(AuthKeys.LOGOUT, authNamespace) logout: any;
  @Getter('isAuthenticated', authNamespace) isAuthenticated!: boolean;

  created() {
    addUnauthorizedInterceptor(this.logout);
  }
}
</script>

<style lang="scss">
#app {
  font-family: "Avenir", Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
}

#nav {
  padding: 30px;

  a {
    font-weight: bold;
    color: #2c3e50;

    &.router-link-exact-active {
      color: #42b983;
    }
  }
}
</style>
