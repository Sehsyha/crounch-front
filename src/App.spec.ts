import { shallowComponent } from './utils/test';
import { createAuthModuleMock } from './store/auth/mockModule';
import App from './App.vue';
import { addUnauthorizedInterceptor } from './api/interceptors';
import { Module } from 'vuex';
import { AuthState } from './store/auth';
import { RootState } from './store';
import { Wrapper } from '@vue/test-utils';
import { AuthGetters } from './store/auth/keys';

jest.mock('./api/interceptors');

describe('App', () => {
  let wrapper: Wrapper<App>;
  let auth: Module<AuthState, RootState>;

  beforeEach(() => {
    auth = createAuthModuleMock();

    const modules = {
      auth,
    };

    wrapper = shallowComponent(App, { modules });
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render when user is authenticated.', () => {
    auth = createAuthModuleMock();
    (auth.getters![AuthGetters.IS_AUTHENTICATED] as jest.Mock).mockReturnValue(true);

    const modules = {
      auth,
    };

    wrapper = shallowComponent(App, { modules });

    expect(wrapper).toMatchSnapshot();
  });

  it('Should add interceptor to api when created.', () => {
    expect(addUnauthorizedInterceptor as jest.Mock).toHaveBeenCalled();
  });
});
