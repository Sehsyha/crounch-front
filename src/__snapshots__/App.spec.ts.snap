// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render. 1`] = `
<div id="app">
  <div id="nav">
    <router-link-stub to="/">Home|</router-link-stub>
    <!---->
    <router-link-stub to="/about">About|</router-link-stub>
    <router-link-stub to="/version">Version</router-link-stub>
  </div>
  <router-view-stub></router-view-stub>
</div>
`;
